library/manageiq.py

```diff
@@ -64,14 +61,86 @@
-    try:
-        result = client.get(providers_url + '/%d/?attributes=authentications,endpoints' % provider_id)
-        endpoint = result['endpoints'][0]
-        return False if (endpoint['hostname'] == hostname and endpoint['port'] == int(port)) else True
-    except Exception as e:
-        msg = "Failed to get provider data. Error: %s" % e
-        module.fail_json(msg=msg)
+class ManageIQ(object):
+    """
+        ManageIQ object to execute various operations in manageiq
+
+        url      - manageiq environment url
+        user     - the username in manageiq
+        password - the user password in manageiq
+    """
+    openshift_provider_type = 'ManageIQ::Providers::OpenshiftEnterprise::ContainerManager'
+
+    def __init__(self, module, url, user, password):
+        self.module        = module
+        self.url           = url
+        self.user          = user
+        self.password      = password
+        self.client        = MiqApi(self.url, (self.user, self.password))
+        self.changed       = False
+        self.msg           = ''
+        self.providers_url = self.url + '/providers'
+
+
+    def update_required(self, provider_id, hostname, port):
+        try:
+            result = self.client.get(self.providers_url + '/%d/?attributes=authentications,endpoints' % provider_id)
+            endpoint = result['endpoints'][0]
+            return False if (endpoint['hostname'] == hostname and endpoint['port'] == int(port)) else True
+        except Exception as e:
+            self.msg = "Failed to get provider data. Error: %s" % e
+            self.module.fail_json(msg=self.msg)
+
+    def update_provider(self, endpoints, provider_name, provider_id):
+        # updates provider with new parameters
+        try:
+            result = self.client.post(self.providers_url + '/%d' % provider_id,
+                                      action='edit',
+                                      connection_configurations=endpoints)
+            self.msg = "Successfuly updated %s provider" % provider_name
+            self.changed = True
+        except Exception as e:
+            self.msg = "Failed to update provider. Error: %s" % e
+            self.module.fail_json(msg=self.msg)
+
+    def add_new_provider(self, provider_name, endpoints):
+        try:
+            result = self.client.post(self.providers_url, type=ManageIQ.openshift_provider_type,
+                                      name=provider_name,
+                                      connection_configurations=endpoints)
+            provider_id = result['results'][0]['id']
+            self.msg = "Successfuly added %s provider" % provider_name
+            self.changed = True
+        except Exception as e:
+            self.msg = "Failed to add provider. Error: %s" % e
+            self.module.fail_json(msg=self.msg)
+        return provider_id
+
+    def add_provider(self, provider_name, hostname, port, token):
```

Comment:
`add_new_provider` and `add_provider` sound more or less equivalent.
Consider `add_or_update_provider` or maybe `set_provider`.